### DISCUSSION TITLE
Drilldown analytics: lazy-loading DuckDB snowflake-schema data into a JS frontend

### DISCUSSION THREAD

**Discussion by ixxie** [duckdb/duckdb#3984] [Q&A]
We are implementing an app with a drilldown analytics interface with a javascript (vue.js) frontend and we are very interested in using DuckDB as it seems to solve a lot of our problems.

However, it is unclear how one key problem is solved with DuckDB: **incremental loading of data to the frontend**. Supposing a snowflake schema, we want to lazy-load rows to the frontend depending on user selection, i.e. depending on the state of the frontend. This sort of thing is [described for example in the documentation of the enterprise version of ag-grid](https://www.ag-grid.com/vue-data-grid/server-side-model/#server-side-row-model).

I had imagined one path to achieving this is storing application state in the backend and writing custom queries to update it, and [some experimental proof-of-concept research project doing this on a WASM client-side database](https://riffle.systems/essays/prelude/); however, I would rather not implement this sort of thing myself, and I imagine there should be existing tools for this in the DuckDB ecosystem.

How would you recommend going about this? Does DuckDB's native API provide a nice way to handle javascript client lazy-loading of a snowflake schema? Does a complementary tool solve this problem?

**Comment by Alex-Monahan** [comment]
Hello!

I have not seen a library that solves your complete problem statement. Very interesting use case though! DuckDB would work very well as a server side pivoting / aggregating / slicing engine. 

Have you seen [DuckDB WASM](https://github.com/duckdb/duckdb-wasm)? You could even do your aggregations client side (with filters pushed down to only pull relevant rows).

**Reply by ixxie** [reply]
I considered that... I am not sure what the tradeoffs between doing everything server-side and synchronizing the client database to the server.... also, not sure exactly how good WASM support is in browsers in the wild.

**Reply by Alex-Monahan** [reply]
WASM support in browsers is excellent! And with DuckDB WASM, you don't need to send the whole DB to the client, just the portions that match the where clause you are using!

**Reply by ixxie** [reply]
I'm worried also about the maintainability of code that performs the lazy loading... this ended up being surprisingly complex as far as we could tell (although perhaps we are missing something).

**Comment by antonycourtney** [comment]
Hi @ixxie,

You may be interested in taking a look at [Tad](https://www.tadviewer.com), an application for viewing and analyzing CSV, Parquet and DuckDb database files. Tad is powered by DuckDb and was [announced in this discussion forum](https://github.com/duckdb/duckdb/discussions/3566) last month.  Tad is an Electron app, and implemented in React and TypeScript.

The core of Tad is a lazily-loaded pivot table. Tad does two types of lazy loading:

1. Tad uses virtual scrolling to only load as much data as needed to construct the current viewport, based on the user's scroll position. Tad will construct a SQL query with `offset` and  `limit` clauses to minimize the amount of data loaded.
2. Tad's pivot table supports lazily loaded drill-down by generating a SQL query to only generate a tree table for the current view based on which nodes the user has expanded in the tree.

Tad isn't really intended to be a general-purpose framework, but the code is pretty modular -- there is a library, `reltab` for composing queries without the need for template strings using a DataFrame-like API, and on top of this is a library, `aggtree`, for constructing the (lazily expanded) pivot tree.  Code is [all on github](https://github.com/antonycourtney/tad).

Maybe not exactly what you're looking for, but I hope that helps a bit!

**Reply by ixxie** [reply]
Interesting! Have you considered separating the lazy-loading engine into a separate package and documenting its API? It would make for a neat framework agnostic way to tackle this problem....
